include/cpp2util.h

```diff
@@ -2084,24 +2087,90 @@ struct args_t
     auto end()    const -> iterator    { return iterator{ argc, argv, argc }; }
     auto cbegin() const -> iterator    { return begin(); }
     auto cend()   const -> iterator    { return end(); }
-    auto size()   const -> std::size_t { return cpp2::unsafe_narrow<std::size_t>(argc); }
+    auto size()   const -> std::size_t { return cpp2::unsafe_narrow<std::size_t>(ssize()); }
     auto ssize()  const -> int         { return argc; }
 
     auto operator[](int i) const {
-        if (0 <= i && i < argc) { return std::string_view{ argv[i] }; }
-        else                    { return std::string_view{}; }
+        if (0 <= i && i < ssize())     { return std::string_view{ argv[i] }; }
+        else                           { return std::string_view{}; }
     }
 
     mutable int        argc = 0;        //  mutable for compatibility with frameworks that take 'int& argc'
     char**             argv = nullptr;
 };
 
-inline auto make_args(int argc, char** argv) -> args_t
+inline auto make_args(int argc, char** argv) -> args
 {
-    return args_t{argc, argv};
+    return args{argc, argv};
 }
 
 
+//-----------------------------------------------------------------------
+//
+//  range: a half-open range of [first, last)
+//
+//-----------------------------------------------------------------------
+//
+template<typename T>
+struct range
+{
+    range(
+        T const& f, 
+        T const& l, 
+        bool include_last = false
+    ) 
+        : first{ f }
+        , last{ l } 
+    { 
+        if (include_last) { 
+            ++last; 
+        }
+    }
+
+    class iterator {
+    public:
+        iterator(T const& f, T const& l, T start) : first{ f }, last{ l }, curr{ start } {}
+
+        auto operator*() const {
+            if (curr != last) { return curr; }
+            else { return T{}; }
+        }
+
+        auto operator+(int i) -> iterator {
+            if (i > 0) { return { first, last, std::min(curr + i, last) }; }
+            else { return { first, last, std::max(curr + i, 0) }; }
+        }
+        auto operator-(int i) -> iterator  { return operator+(-i); }
+        auto operator++()     -> iterator& { if (curr != last ) { ++curr; }  return *this; }
+        auto operator--()     -> iterator& { if (curr != first) { --curr; }  return *this; }
+        auto operator++(int)  -> iterator  { auto old = *this;  ++*this;  return old; }
+        auto operator--(int)  -> iterator  { auto old = *this;  ++*this;  return old; }
+
+        auto operator<=>(iterator const&) const = default;
+
+    private:
+        T first;
+        T last;
+        T curr;
+    };
+
+    auto begin()  const -> iterator { return iterator{ first, last, first }; }
+    auto end()    const -> iterator { return iterator{ first, last, last }; }
+    auto cbegin() const -> iterator { return begin(); }
+    auto cend()   const -> iterator { return end(); }
+    auto size()   const -> std::size_t { return /*cpp2::unsafe_narrow<std::size_t>*/(ssize()); }
```

Comment:
[Question] Do you want to eventually remove the commented part or keep it aligned with the `args::iterator` code in line 2090?

Reply:
Thanks, good catch -- I had commented that out while debugging something, which I guess I've now fixed so I can restore that code. 👍